### PR TITLE
[stable27] 44032 fix show new shares without refresh

### PR DESCRIPTION
**Pull request by nfebe**
Resolves : https://github.com/nextcloud/server/issues/44032

#### Before

https://github.com/nextcloud/server/assets/14317775/565747a6-4e84-448f-9730-d08ffe606f72 

#### After

https://github.com/nextcloud/server/assets/14317775/aca51792-fad2-47cb-ac03-ccd460a02c0a

### CODE REVIEW

File: apps/files_sharing/src/components/SharingEntryLink.vue [hunk 1]
```diff
@@ -57,8 +57,7 @@
 			class="sharing-entry__actions"
 			:aria-label="actionsTooltip"
 			menu-align="right"
-			:open.sync="open"
-			@close="onCancel">
```
Comment (artonge) [inline apps/files_sharing/src/components/SharingEntryLink.vue:61]: Is the pending share removed if the user clicks away?

Comment (Fenn-CS) [reply apps/files_sharing/src/components/SharingEntryLink.vue:61]: We explicitly call cancel, when the cancel button is pressed, if click-outside, it closes the menu without doing anything, new share processes starts with a new share object.

Comment (artonge) [reply apps/files_sharing/src/components/SharingEntryLink.vue:61]: If you create a second public link, then clicking away do not remove the pending share. It does when clicking "Cancel". I think clicking away should have the same behavior as canceling.

File: apps/files_sharing/src/components/SharingEntryLink.vue [hunk 2]
```diff
@@ -407,6 +408,29 @@ export default {
 			return this.config.isDefaultExpireDateEnforced && this.share && !this.share.id
 		},
 
+		sharePolicyHasRequiredProperties() {
+			return this.config.enforcePasswordForPublicLink || this.config.isDefaultExpireDateEnforced
+		},
+
+		requiredPropertiesMissing() {
+			// Destructure share and config from this
+			const { share, config } = this
```
Comment (susnux) [inline apps/files_sharing/src/components/SharingEntryLink.vue:417]: Why is this needed? Never saw this before in our code base, so maybe just directly access this.share

File: apps/files_sharing/src/components/SharingEntryLink.vue [hunk 3]
```diff
@@ -407,6 +408,29 @@ export default {
...
+			// Ensure share exist and the share policy has required properties
+			if (!this.sharePolicyHasRequiredProperties) return false
+
+			if (share) {
+				// If share has ID, then this is an incoming link share created from the existing link share
+				// Hence assume required properties
+				if (share.id) return true
```
Comment (susnux) [inline apps/files_sharing/src/components/SharingEntryLink.vue:424]:
```suggestion
				if (share.id) {
					return true
				}
```

We never use short syntax.

File: apps/files_sharing/src/components/SharingEntryLink.vue [hunk 4]
```diff
@@ -407,6 +408,29 @@ export default {
...
+				// Check if either password or expiration date is missing and enforced
+				const isPasswordMissing = config.enforcePasswordForPublicLink && (share.password === null || share.password === undefined)
```
Comment (susnux) [inline apps/files_sharing/src/components/SharingEntryLink.vue:426]: Can we not use
```suggestion
				const isPasswordMissing = config.enforcePasswordForPublicLink && !share.password
```

?

File: apps/files_sharing/src/components/SharingEntryLink.vue [hunk 5]
```diff
@@ -483,6 +507,7 @@ export default {
 		 * Create a new share link and append it to the list
 		 */
 		async onNewLinkShare() {
+			console.debug('onNewLinkShare called (this.share status)', this.share)
```
Comment (susnux) [inline apps/files_sharing/src/components/SharingEntryLink.vue:510]: Please use the logger. The logger respects the configured logging level. 

```suggestion
			this.logger.debug('onNewLinkShare called (this.share status)', { share: this.share })
```

File: apps/files_sharing/src/components/SharingEntryLink.vue [hunk 6]
```diff
@@ -497,28 +522,13 @@ export default {
 				shareDefaults.expiration = this.formatDateToString(this.config.defaultExpirationDate)
 			}
 
+			console.debug('Missing required properties?', this.requiredPropertiesMissing)
```
Comment (susnux) [inline apps/files_sharing/src/components/SharingEntryLink.vue:525]: Same

File: apps/files_sharing/src/components/SharingEntryLink.vue [hunk 7]
```diff
@@ -540,8 +550,33 @@ export default {
...
+					// if the share is valid, create it on the server
+					if (this.checkShare(this.share)) {
+						try {
+							this.logger.info('Sending existing share to server', this.share)
+							await this.pushNewLinkShare(this.share, true)
+							this.shareCreationComplete = true
+							this.logger.info('Share created on server', this.share)
+						} catch (e) {
+							this.pending = false
+							console.error(e)
```
Comment (susnux) [inline apps/files_sharing/src/components/SharingEntryLink.vue:565]: Same (logger)

File: apps/files_sharing/src/components/SharingEntryLink.vue [hunk 8]
```diff
@@ -540,8 +550,33 @@ export default {
...
+							this.logger.error('Error creating share', e)
+							return false
+						}
+						return true
+					} else {
+						this.open = true
+						OC.Notification.showTemporary(t('files_sharing', 'Error, please enter proper password and/or expiration date'))
```
Comment (susnux) [inline apps/files_sharing/src/components/SharingEntryLink.vue:572]: Better use `showError`

File: apps/files_sharing/src/components/SharingEntryLink.vue [hunk 9]
```diff
@@ -752,25 +791,23 @@ export default {
...
+				overflow: hidden;
+				white-space: nowrap;
+			}
 		}
-	  }
-
-	  &__copy {
 
-	  }
+		&__copy {}
```
Comment (susnux) [inline apps/files_sharing/src/components/SharingEntryLink.vue:810]: Empty rule?

**Review by susnux — COMMENTED** [review]
Logically makes sense, some code comments.

File: apps/files_sharing/src/components/SharingEntryLink.vue (hunk 2 above)
Comment (Fenn-CS) [reply apps/files_sharing/src/components/SharingEntryLink.vue:417]: The idea was to shorten these lines a bit

```js
const isPasswordMissing = config.enforcePasswordForPublicLink && (share.password === null || share.password === undefined)
const isExpireDateMissing = config.isDefaultExpireDateEnforced && (share.expireDate === null || share.expireDate === undefined)
```

File: apps/files_sharing/src/components/SharingEntryLink.vue (hunk 9 above)
Comment (Fenn-CS) [reply apps/files_sharing/src/components/SharingEntryLink.vue:810]: Lint fix formatting, was probably already emtpy

File: apps/files_sharing/src/components/SharingEntryLink.vue (hunk 2 above)
Comment (Fenn-CS) [reply apps/files_sharing/src/components/SharingEntryLink.vue:417]: The shortened version you proposed below, would suffice, so can take this out.

**Review by susnux — APPROVED** [review]
Code looks good now

File: apps/files_sharing/src/mixins/SharesMixin.js [hunk 10]
```diff
@@ -80,6 +82,10 @@ export default {
 			 * ! do not remove it ot you'll lose all reactivity here
 			 */
 			reactiveState: this.share?.state,
+			logger: getLoggerBuilder()
```
Comment (artonge) [inline apps/files_sharing/src/mixins/SharesMixin.js:85]: I think this should better be an imported service rather than in this mixin.

File: apps/files_sharing/src/components/SharingEntryLink.vue [hunk 11]
```diff
@@ -407,6 +408,31 @@ export default {
...
+					return true
+				}
+				// Check if either password or expiration date is missing and enforced
+				const isPasswordMissing = this.config.enforcePasswordForPublicLink && !this.share.password
+				const isExpireDateMissing = this.config.isDefaultExpireDateEnforced && !this.share.expireDate
+
+				return isPasswordMissing || isExpireDateMissing
+			}
+			// if no share, we can't tell if properties are missing or not so we assume properties are missing
+			return true
```
Comment (artonge) [inline apps/files_sharing/src/components/SharingEntryLink.vue:434]: Maybe you can return early for that one.

**Review by artonge — APPROVED** [review]
Looks good :)
